sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -524,14 +635,150 @@ protected EventProcessor()
+                    }
+                }
+                catch (Exception ex) when ((ex is OperationCanceledException) || (ex is TaskCanceledException))
+                {
+                    // This is expected as part of the normal flow; no action is needed.
+                }
+                catch (Exception ex)
+                {
+                    Logger.EventProcessingTaskError(Identifier, EventHubName, ConsumerGroup, ex.Message);
+                    processingException = ex;
+                }
+
+                // With the processing task having completed, perform the necessary cleanup of partition processing tasks
+                // and surrender ownership.
+
+                var stopPartitionProcessingTasks = ActivePartitionProcessors.Keys
+                    .Select(partitionId => StopProcessingPartitionAsync(partitionId, ProcessingStoppedReason.Shutdown, CancellationToken.None))
+                    .ToArray();
+
+                if (async)
+                {
+                    await Task.WhenAll(stopPartitionProcessingTasks).ConfigureAwait(false);
+                    await LoadBalancer.RelinquishOwnershipAsync(cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    Task.WaitAll(stopPartitionProcessingTasks);
+
+#pragma warning disable AZC0102 // Do not use GetAwaiter().GetResult(). Use the TaskExtensions.EnsureCompleted() extension method instead.
+                    LoadBalancer.RelinquishOwnershipAsync(cancellationToken).GetAwaiter().GetResult();
+#pragma warning restore AZC0102 // Do not use GetAwaiter().GetResult(). Use the TaskExtensions.EnsureCompleted() extension method instead.
+                }
+
+                // Dispose of the processing task and reset processing state to
+                // allow the processor to be restarted when this method completes.
+
+                _runningProcessorTask.Dispose();
+                _runningProcessorTask = null;
+            }
+            catch (Exception ex)
+            {
+                Logger.EventProcessorStopError(Identifier, EventHubName, ConsumerGroup, ex.Message);
+                throw;
             }
+            finally
+            {
+                Logger.EventProcessorStopComplete(Identifier, EventHubName, ConsumerGroup);
+
+                // If the cancellation token was signaled during the attempt to acquire the
+                // semaphore, it cannot be safely released; ensure that it is held.
+
+                if (releaseSync)
+                {
+                    ProcessorRunningSync.Release();
+                }
+            }
+
+            // Surface any exception that was captured when the processing task was
+            // initially awaited.
 
+            if (processingException != default)
+            {
+                throw processingException;
```

Comment:
Wondering if we should use [`ExceptionDispatchInfo`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.exceptionservices.exceptiondispatchinfo) in some way here.  I think as is this is going to cause the existing stack trace from `processingException` to be overwritten, so we'll never be able to see where inside the guts of the processor the fault was.

Reply:
Good call.  I forgot about the dispatcher.  The exception will already have been surfaced within the event handler with the original stack trace.  but we should try to preserve it here as well.